### PR TITLE
(feat) add prerequisite_tutorials capability

### LINKED ISSUE #1945: Add Capability to List Tutorial Prerequisites as Meta Data
Currently, some tutorials list prerequisites in their introduction section, and some don't, and listing (if present) varies. This is not optimal for user experience and maintenance.

1. Add a field `prerequisites_tutorials : string list` to both the `metadata` and `t` types in `tutorial.ml`. This holds the slugs of all the tutorials that are required reading _before_ this one.
1. Add some sensible references between tutorials in the `data/tutorials` folder for testing purposes (for instance, in “Values and Function” and “Basic Data Types and Pattern Matching”)
1. In `handler.ml`, in the function `tutorial`, you need to filter the tutorials for those that are prerequisites, i.e. (`slug` is an element of `tutorial.prerequisite_tutorials`).
1. In `tutorial.eml`, render the list of links to the prerequisite tutorials as a paragraph at the beginning.

Note: The fields `prerequisites_tutorials` cannot be computed from the `recommended_next_tutorials` defined in issue #1589 and implemented in PR #1928. Prerequisites are mandatory, but suggested next are optional.

### PR DESCRIPTION
**Pull request by PoorlyDefinedBehaviour**
Here's how it looks like at the moment:
image

Made the field an option because I get this error otherwise:
```
OCAMLORG_HTTP_PORT=8000 make start
opam exec -- dune build --root .
File "src/ocamlorg_data/dune", line 98, characters 0-126:
 98 | (rule
 99 |  (target tutorial.ml)
100 |  (action
101 |   (with-stdout-to
102 |    %{target}
103 |    (run %{dep:../../tool/ood-gen/bin/gen.exe} tutorial))))
ood-gen: internal error, uncaught exception:
         Ood_gen.Exn.Decode_error("Didn't find the function for key: prerequisites_tutorials")
         Raised at Ocamlorg__Import.Result.get_ok.(fun) in file "src/global/import.ml", line 103, characters 55-70
         Called from Ood_gen__Tutorial.all in file "tool/ood-gen/lib/tutorial.ml", line 136, characters 2-41
         Called from Ood_gen__Tutorial.template in file "tool/ood-gen/lib/tutorial.ml", line 184, characters 4-12
         Called from Dune__exe__Gen.cmds.(fun) in file "tool/ood-gen/bin/gen.ml", line 39, characters 48-61
         Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
         Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 34, characters 37-44
make: *** [Makefile:5: all] Error 1      
```

Closes https://github.com/ocaml/ocaml.org/issues/1945

### CODE REVIEW

File: src/ocamlorg_frontend/pages/tutorial.eml
Comment (cuihtlauac) [inline src/ocamlorg_frontend/pages/tutorial.eml:1]: Rendering is ok, although the font looks bigger than other `<h2> / ##` header.

We need to find a way to have the prerequisite list _after_ the title, as I suggested, it looks clumsy. Sorry, I hadn't realized that issue when writing the issue.

File: data/tutorials/language/0it_01_basic_datatypes.md [hunk 1]
```diff
@@ -4,6 +4,11 @@ title: Basic Data Types and Pattern Matching
 description: |
   Predefined Types, Variants, Records, and Pattern Matching
 category: "Introduction"
+prerequisites_tutorials:
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_01_basic_datatypes.md]:
```suggestion
prerequisite_tutorials: 
```

File: data/tutorials/language/0it_04_labels.md [hunk 2]
```diff
@@ -4,6 +4,10 @@ title: Labelled and Optional Arguments
 description: >
   Provide labels to your functions arguments
 category: "Introduction"
+prerequisites_tutorials:
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_04_labels.md]:
```suggestion
prerequisite_tutorials: 
```

File: src/ocamlorg_data/data.mli [hunk 3]
```diff
@@ -235,6 +236,7 @@ module Tutorial : sig
     toc : toc list;
     body_html : string;
     recommended_next_tutorials : recommended_next_tutorials option;
+    prerequisites_tutorials : prerequisites_tutorials option;
```
Comment (cuihtlauac) [inline src/ocamlorg_data/data.mli]:
```suggestion
    prerequisite_tutorials : prerequisites_tutorials option;
```

File: src/ocamlorg_frontend/pages/tutorial.eml [hunk 4]
```diff
@@ -37,11 +37,27 @@ let render_related_exercises exercises =
   else
     ""
 
+let render_prerequisites_tutorials tutorials =
```
Comment (cuihtlauac) [inline src/ocamlorg_frontend/pages/tutorial.eml]:
```suggestion
let render_prerequisite_tutorials tutorials =
```

File: tool/ood-gen/lib/tutorial.ml [hunk 5]
```diff
@@ -29,13 +29,17 @@ type external_tutorial = {
 type recommended_next_tutorials = string list
 [@@deriving of_yaml, show { with_path = false }]
 
+type prerequisites_tutorials = string list
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/tutorial.ml]:
```suggestion
type prerequisite_tutorials = string list
```

File: tool/ood-gen/lib/tutorial.ml [hunk 6]
```diff
@@ -29,13 +29,17 @@ type external_tutorial = {
...
+[@@deriving of_yaml, show { with_path = false }]
+
 type metadata = {
   id : string;
   title : string;
   description : string;
   category : string;
   external_tutorial : external_tutorial option;
   recommended_next_tutorials : recommended_next_tutorials option;
+  prerequisites_tutorials : prerequisites_tutorials option;
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/tutorial.ml]:
```suggestion
  prerequisites_utorials : prerequisite_tutorials option;
```

File: tool/ood-gen/lib/tutorial.ml [hunk 7]
```diff
@@ -51,6 +55,7 @@ type t = {
   body_md : string;
   body_html : string;
   recommended_next_tutorials : recommended_next_tutorials option;
+  prerequisites_tutorials : prerequisites_tutorials option;
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/tutorial.ml]:
```suggestion
  prerequisite_tutorials : prerequisite_tutorials option;
```

File: tool/ood-gen/lib/tutorial.ml [hunk 8]
```diff
@@ -157,6 +162,7 @@ type external_tutorial =
   ; contribute_link : contribute_link
   }
 type recommended_next_tutorials = string list
+type prerequisites_tutorials = string list
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/tutorial.ml]:
```suggestion
type prerequisite_tutorials = string list
```

File: tool/ood-gen/lib/tutorial.ml [hunk 9]
```diff
@@ -169,6 +175,7 @@ type t =
   ; toc : toc list
   ; body_html : string
   ; recommended_next_tutorials : recommended_next_tutorials option
+  ; prerequisites_tutorials : prerequisites_tutorials option
```
Comment (cuihtlauac) [inline tool/ood-gen/lib/tutorial.ml]:
```suggestion
  ; prerequisite_tutorials : prerequisite_tutorials option
```

File: src/ocamlorg_web/lib/handler.ml [hunk 10]
```diff
@@ -409,10 +409,19 @@ let tutorial req =
...
+      (Option.value tutorial.prerequisites_tutorials ~default:[])
+  in
+
+  let prerequisites_tutorials = all_tutorials |> List.filter is_prerequisite in
+
   Dream.html
     (Ocamlorg_frontend.tutorial ~tutorials
        ~canonical:(Url.tutorial tutorial.slug)
-       ~related_exercises ~recommended_next_tutorials tutorial)
+       ~related_exercises ~recommended_next_tutorials ~prerequisites_tutorials
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
       ~related_exercises ~recommended_next_tutorials ~prerequisite_tutorials
```

File: src/ocamlorg_web/lib/handler.ml [hunk 11]
```diff
@@ -409,10 +409,19 @@ let tutorial req =
     all_tutorials |> List.filter is_in_recommended_next
   in
 
+  let is_prerequisite (tested : Data.Tutorial.t) =
+    List.exists
+      (fun r -> r = tested.slug)
+      (Option.value tutorial.prerequisites_tutorials ~default:[])
```
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml]:
```suggestion
      (Option.value tutorial.prerequisite_tutorials ~default:[])
```

File: src/ocamlorg_web/lib/handler.ml
Comment (cuihtlauac) [inline src/ocamlorg_web/lib/handler.ml:1]: Parsing tutorial's Yaml header should not produce an `option list` but a `list`. The absence of the `prerequisite_tutorials` header should result in an empty list; wrapping in an option is redundant.

I'd also prefer having `List.map` over the `prerequisite_tutorials` `string list` to turn it into a `Data.Tutorial.t list`.

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Thanks, @PoorlyDefinedBehaviour, I've left many comments, but this is great.

**Comment by cuihtlauac** [comment]
Regarding the typing `prerequisite_tutorials` as a `list` instead of `list option`, you can have a look at PR #1967, it does it for recommended next tutorials.

**Review by christinerose — APPROVED** [review]

File: src/ocamlorg_frontend/pages/tutorial.eml
Comment (PoorlyDefinedBehaviour) [reply src/ocamlorg_frontend/pages/tutorial.eml:1]: I'll look into it.

**Comment by cuihtlauac** [comment]
This looks good to me. However, we need to resolve the title and prerequisites placement issue before merging. Once issue #1971 is resolved, it should be easier to solve this one.

**Review by christinerose — APPROVED** [review]

**Comment by cuihtlauac** [comment]
@PoorlyDefinedBehaviour Can you rebase and do the changes so we can merge this?

**Comment by PoorlyDefinedBehaviour** [comment]
> @PoorlyDefinedBehaviour Can you rebase and do the changes so we can merge this?

Done

**Comment by cuihtlauac** [comment]
Thanks, @PoorlyDefinedBehaviour; this is looking great. Now, we need to remove the old prerequisites paragraph in the introductions. This needs to be done with care in order to remove only prerequisites not unrelated text.

File: data/tutorials/language/0it_04_labels.md [hunk 12]
```diff
@@ -5,14 +5,16 @@ short_title: Labelled and Optional Arguments
 description: >
   Provide labels to your functions arguments
 category: "Introduction"
+prerequisite_tutorials: 
+  [
+    "values-and-functions"
+  ]
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_04_labels.md]:
```suggestion
prerequisite_tutorials:
  - "values-and-functions"
```

Let's use the Yaml item list syntax

File: data/tutorials/language/0it_01_basic_datatypes.md
Comment (cuihtlauac) [inline data/tutorials/language/0it_01_basic_datatypes.md:1]: l 23 should be turned into something like:
```md
**Note**: As in previous tutorials, expressions after `#` and ending with `;;` are for the toplevel, like UTop.
```

File: data/tutorials/language/0it_01_basic_datatypes.md [hunk 13]
```diff
@@ -5,6 +5,11 @@ short_title: Basic Data Types and Pattern Matching
 description: |
   Predefined Types, Variants, Records, and Pattern Matching
 category: "Introduction"
+prerequisite_tutorials: 
+  [
+    "tour-of-ocaml",
+    "values-and-functions",
+  ]
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_01_basic_datatypes.md]:
```suggestion
  - tour-of-ocaml"
  - "values-and-functions"
```

**Review by cuihtlauac — CHANGES_REQUESTED** [review]

File: data/tutorials/language/0it_01_basic_datatypes.md
Comment (PoorlyDefinedBehaviour) [reply data/tutorials/language/0it_01_basic_datatypes.md:1]: Done

**Review by cuihtlauac — APPROVED** [review]
This looks good to me. Many thanks, @PoorlyDefinedBehaviour, this is brilliant!

**Review by sabine — APPROVED** [review]
Merging! :sparkles: :+1: Thanks @PoorlyDefinedBehaviour !

**Pull request merged**